src/pymor/operators/constructions.py

```diff
@@ -1454,3 +1454,74 @@ def apply(self, U, mu=None):
 
     def as_range_array(self, mu=None):
         return self.B.as_range_array(mu).lincomb(mu['input'])
+
+
+class QuadraticFunctional(Operator):
+    """An `Operator` representing a quadratic functional.
+
+    Given an operator `A` acting on R^d, this class represents the functional::
+
+        op: R^d ----> R^1
+             u  |---> (A(u), u).
+
+    Parameters
+    ----------
+    operator
+        The |Operator| defining the quadratic functional.
+    """
+
+    linear = False
+    range = NumpyVectorSpace(1)
+
+    def __init__(self, operator, name=None):
+        assert operator.linear
+        assert operator.source == operator.range
+        self.__auto_init(locals())
+        self.source = operator.source
+
+    def apply(self, U, mu=None):
+        assert U in self.source
+        return self.range.from_numpy(self.operator.apply2(U, U, mu))
+
+    def jacobian(self, U, mu=None):
+        inner_vec = self.operator.apply_adjoint(U, mu) + self.operator.apply(U, mu)
+        return VectorFunctional(inner_vec, name=self.name + '_jacobian')
+
+    def d_mu(self, parameter, index=1):
+        # the parameter derivative only takes effect on the inner operator
+        return QuadraticFunctional(self.operator.d_mu(parameter, index), name=self.name + '_d_mu')
+
+
+class QuadraticProductFunctional(QuadraticFunctional):
+    """An `Operator` representing a quadratic functional by multiplying two linear functionals.
+
+    Given linear operators `A`, `B`: R^d ----> R^n, this class represents the functional::
+
+        op: R^d ----> R^1
+             u  |---> (A(u), B(u)).
+
+    Parameters
+    ----------
+    left
+        The |Operator| that defines the left operator of the quadratic functional.
+    right
+        The |Operator| that defines the right operator of the quadratic functional.
+    product
```

Comment:
Do you really have a use case for the product operator here? You assume that the range is one-dimensional, so the product would just be multiplication with a number. If there is no good reason to have it, I would take it out.

Reply:
I don't have a specific example in mind, but the ranges of the individual operators are never assumed to be 1D - the range of the entire product operator is.

Reply:
Ah, sorry, I was confused. Forget my comment.